bridge/matrix/matrix.go

```diff
@@ -216,6 +225,73 @@ func (b *Bmatrix) handlematrix() {
 	}()
 }
 
+func (b *Bmatrix) setDisplayName(mxid string, displayName string) string {
+	now := time.Now()
+
+	// scan to delete old entries, to stop memory usage from becoming too high with old entries
+	toDelete := []string{}
+	b.RLock()
+	for k, v := range b.NicknameMap {
+		if now.Sub(v.lastUpdated) > 10*time.Minute {
+			toDelete = append(toDelete, k)
+		}
+	}
+	b.RUnlock()
+
+	b.Lock()
+	if len(toDelete) != 0 {
+		for _, v := range toDelete {
+			delete(b.NicknameMap, v)
+		}
+	}
+	b.NicknameMap[mxid] = NicknameCacheEntry{
+		displayName: displayName,
+		lastUpdated: now,
+	}
+	b.Unlock()
+
+	return displayName
+}
+
+func (b *Bmatrix) getDisplayName(mxid string) string {
+	if b.GetBool("UseUserName") {
+		return mxid[1:]
+	}
+
+	b.RLock()
+	if val, present := b.NicknameMap[mxid]; present {
+		if time.Since(val.lastUpdated) < 10*time.Minute {
+			b.RUnlock()
+
+			return val.displayName
+		}
+	}
+	b.RUnlock()
+
+	displayName, err := b.mc.GetDisplayName(mxid)
+	var httpError *matrix.HTTPError
+	if errors.As(err, &httpError) {
+		b.Log.Warnf("Couldn't retrieve the display name for %s", mxid)
+	}
+
+	if err != nil {
+		return b.setDisplayName(mxid, mxid[1:])
+	}
+
+	return b.setDisplayName(mxid, displayName.DisplayName)
+}
+
+func (b *Bmatrix) handleMemberChange(ev *matrix.Event) {
+	// Update the displayname on join messages, according to https://matrix.org/docs/spec/client_server/r0.6.1#events-on-change-of-profile-information
+	if ev.Content["membership"] == "join" {
+		dn, dnPresent := ev.Content["displayname"]
+		dnStr, dnValidString := dn.(string)
+		if dnPresent && dnValidString {
+			b.setDisplayName(ev.Sender, dnStr)
+		}
```

Comment:
not tested, but shouldn't this work also?

```golang
if dn,ok := ev.Content["displayname"].(string);ok {
    b.setDisplayName(ev.Sender, dn)
}
```

Reply:
Oh, I didn't know one could chain the presence check and the cast check like, that's nice ;)